### ISSUE TITLE
3.7: Plugins can no longer register settings pages during `admin_init`

### ISSUE THREAD

**Issue by ChaseWiseman** [woocommerce/woocommerce#24389]
**Describe the bug**
As of v3.7, plugins aren't able to add their own settings pages using the `woocommerce_get_settings_pages` filter if the code being used to hook into that filter is init by a filter later than `wp_loaded`. This seems to be caused by the changes in https://github.com/woocommerce/woocommerce/commit/fbe1bc7c1f29b9e321d7976fcd3788a960bb556d

The problem stems from the fact that `WC_Admin_Settings::get_settings_pages()` only applies the `woocommerce_get_settings_pages` filter once on the first call. Subsequent calls will bypass the filter and use what's stored in `WC_Admin_Settings::$settings`. Since `wp_loaded` is now the first touch point for that method, anything hooking into that filter later won't be applied.

Some plugins use `admin_init` to instantiate their admin includes & handlers, so as of 3.7 those plugins' handler's use of `woocommerce_get_settings_pages` will go uncalled.

While it's fairly simple to work around by changing the custom code/plugin's loading, I'm reporting here because it's a change from previous behavior and can cause settings pages to disappear without intervention.

**To Reproduce**
Barebones example, but suppose you have some code like so:
```
function init_great_admin() {

	// other great admin stuff

	// this is probably in a class constructor or something similar
	add_filter( 'woocommerce_get_settings_pages', function( $pages ) {

		$pages[] = include( 'great-settings-page.php' );

		return $pages;

	} );
}
add_action( 'admin_init', 'init_great_admin' );
```

and `great-settings-page.php` contains a settings page class:
```
class Great_Settings_Page extends \WC_Settings_Page {


	public function __construct() {

		$this->id               = 'great-settings';
		$this->label            = 'Great Settings';

		parent::__construct();
	}


	public function output() {

		echo "We're so great!";
	}
}

new Great_Settings_Page();
```

* In WC < 3.7, the custom settings tab and page are loaded
* In WC 3.7, they are not

**Expected behavior**
Consistent with 3.6.5 and below. 

**Isolating the problem (mark completed items with an [x]):**
- [x] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [x] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/storefront/).
- [x] I can reproduce this bug consistently using the steps above.

**Comment by kloon** [comment]
We have had a bit of a discussion around this issue and while programmatically it might not be wrong to attach a filter withing another hook, nowhere do we show this as an example. of how to create settings pages that way, In fact, we had a look through our extension codebase and all our extensions simply attach the filter in the constructor with an is_admin check to prevent it from running on non-admin pages.

So in conclusion, while it was something that worked previously it is not something we really supported.

I am open to hearing arguments for having this work on the admin_init hooks should there be any.

cc @woocommerce/proton
